[Docs] High Availability Serve Controller (#5472)
skypilot-org/skypilot: docs/source/serving/high-availability.rst

.. _skyserve-high-availability-controller:

=================================
High Availability SkyServe Controller
=================================

Overview
--------
By default, the SkyServe controller runs as a single instance (either a VM or a Kubernetes Pod). If this instance fails due to node issues, pod crashes, or other unexpected events, the service endpoint becomes unavailable until the controller is manually recovered or relaunched.

To enhance resilience and ensure service continuity, SkyServe offers a High Availability (HA) mode for its controller. When enabled, the controller leverages Kubernetes capabilities to automatically recover from failures.

> Benefits of HA Controller:

Reviewer: Use sentence case ("Benefits of HA controller", "How to enable HA mode", etc)
There should be a lint check for this - if you push a new commit hopefully the CI will trigger. Maybe it got confused because of the branch names.

Author: Good catch! Thanks for reminding.

~~~~~~~~~~~~~~~~~~~~~~~~~~
* **Automatic Recovery:** The controller and load balancer can automatically restart after crashes or node failures.
* **Service Continuity:** Minimizes downtime for your served applications.
* **Managed by Kubernetes:** Relies on robust, industry-standard Kubernetes features for state management and process supervision.

Prerequisites
-------------
* **Kubernetes Cluster:** High Availability mode is **currently only supported when using Kubernetes** as the cloud provider for the SkyServe controller. You must have a Kubernetes cluster configured for SkyPilot. See :ref:`Kubernetes Setup <kubernetes-setup>` for details.
* **Persistent Kubernetes:** The underlying Kubernetes cluster (control plane and nodes) must be running persistently. If using a local Kubernetes deployment (e.g., Minikube, Kind via `sky local up`), the machine hosting the cluster must remain online.

.. note::
    Currently, HA mode is only supported for Kubernetes. Support for other clouds (e.g., AWS, GCP, Azure VMs) is under development.

How to Enable HA Mode
---------------------

> To enable High Availability for the SkyServe controller, set the ``high_availability`` flag to ``true`` within the ``serve.controller`` section of your SkyPilot configuration file (``~/.sky/config.yaml``):

Reviewer: ```suggestion
To enable High Availability for the SkyServe controller, set the ``high_availability`` flag to ``true`` within the ``serve.controller`` section of your :ref:`SkyPilot configuration <config-yaml>`:
```

.. code-block:: yaml
    :emphasize-lines: 8

    serve:
      # NOTE: these settings only take effect for a *new* SkyServe controller,
      # not if you have an existing one (even if stopped).
      controller:
        # --- Controller Resources ---
        # Optional: Specify resources like cloud, region, cpus, disk_size
        resources:
          cloud: kubernetes  # HA mode requires Kubernetes
          # region: <your-k8s-region> # Optional
          cpus: 2+           # Optional, example value
          # disk_size: 100     # Optional, example value (affects PVC size in HA mode)

        # --- Enable High Availability ---

>         high_availability: true

Reviewer: Can we add this all to `reference/config.rst` as well?

Author: SkyServe config are not added to `config.rst`. All stored seperately in https://docs.skypilot.co/en/latest/serving/sky-serve.html#customizing-skyserve-controller-resources.

Reviewer: We have a similar detailed breakdown for managed jobs https://docs.skypilot.co/en/latest/examples/managed-jobs.html#customizing-jobs-controller-resources but I think we should still include all the possible configuration in `config.rst` for completeness. `config.rst` can link back here for details.

.. note::

>     Enabling or disabling ``high_availability`` only affects **new** SkyServe controllers. If you have an existing controller (either running or stopped), changing this setting will not modify it. To apply the change, you must first terminate all services and then tear down the existing controller using ``sky down --purge <controller_name>``. See `Important Considerations`_ below.

Reviewer: Is `--purge` needed?

Author: Yeah it's dangerous. Let's remove this flag.

Configuration Details
---------------------
Besides the main ``serve.controller.high_availability: true`` flag, you can customize HA behavior further:

* **Controller Resources (``serve.controller.resources``):** As usual, you can specify ``cloud`` (must be Kubernetes), ``region``, ``cpus``, etc. The ``disk_size`` here directly determines the size of the PersistentVolumeClaim created for the HA controller.
* **Kubernetes Storage Class (``kubernetes.high_availability.storage_class_name`` - Optional):** If your Kubernetes cluster has specific storage classes defined (e.g., for different performance tiers like SSD vs HDD, or specific features like backup), you can specify which one to use for the controller's PVC. This is configured under the ``kubernetes`` section in ``config.yaml``:

    .. code-block:: yaml

        kubernetes:
          # ... other kubernetes settings ...
          high_availability:
            # Optional: Specify the StorageClass name for the controller's PVC

>             storage_class_name: <your-storage-class-name> # e.g., premium-ssd

Reviewer: same here about adding to config.rst

.. _skyserve-high-availability-controller:

=========================================
High Availability SkyServe Controller
=========================================

Overview
--------
By default, the SkyServe controller runs as a single instance (either a VM or a Kubernetes Pod). If this instance fails due to node issues, pod crashes, or other unexpected events, the service endpoint becomes unavailable until the controller is manually recovered or relaunched.

To enhance resilience and ensure service continuity, SkyServe offers a High Availability mode for its controller. When enabled, the controller leverages the automatic recovery feature of Kubernetes Deployments to automatically recover from failures.

Benefits of high availability controller:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
* **Automatic Recovery:** The controller and load balancer can automatically restart after crashes or node failures.
* **Service Continuity:** Minimizes downtime for your served applications.

> * **Managed by Kubernetes:** Relies on robust, industry-standard Kubernetes features for state management and process supervision.

Reviewer 2: We can skip this? Seems not a benefit?

Prerequisites
-------------
* **Kubernetes Cluster:** High Availability mode is **currently only supported when using Kubernetes** as the cloud provider for the SkyServe controller. You must have a Kubernetes cluster configured for SkyPilot. See :ref:`Kubernetes Setup <kubernetes-setup>` for details.
* **Persistent Kubernetes:** The underlying Kubernetes cluster (control plane and nodes) must be running persistently. If using a local Kubernetes deployment (e.g., Minikube, Kind via ``sky local up``), the machine hosting the cluster must remain online.

> 

Reviewer 2: Need PVC support as well?

.. note::
    Currently, High Availability mode is only supported for Kubernetes. Support for other clouds (e.g., AWS, GCP, Azure VMs) is under development.

How to enable high availability mode
-------------------------------------

> To enable High Availability for the SkyServe controller, set the ``high_availability`` flag to ``true`` within the ``serve.controller`` section of your :ref:`SkyPilot configuration <config-yaml>`:

Reviewer 2: How about we lower case `High Availability` everywhere?

Author: Good point! Let's use lower cases for consistency.

.. code-block:: yaml
    :emphasize-lines: 8,11

    serve:
      # NOTE: these settings only take effect for a *new* SkyServe controller,
      # not if you have an existing one (even if stopped).
      controller:
        # --- Controller Resources ---
        # Optional: Specify resources like cloud, region, cpus, disk_size
        resources:
          cloud: kubernetes  # High Availability mode requires Kubernetes
          # region: <your-k8s-region> # Optional
          cpus: 2+           # Optional, example value
          # disk_size: 100     # Optional, example value (affects PVC size in High Availability mode)

        # --- Enable High Availability ---

>         high_availability: true

Reviewer 2: Add serve field in our advanced configuration doc as well? https://docs.skypilot.co/en/docs-serve-ha/reference/config.html

.. note::
    Enabling or disabling ``high_availability`` only affects **new** SkyServe controllers. If you have an existing controller (either running or stopped), changing this setting will not modify it. To apply the change, you must first terminate all services and then tear down the existing controller using ``sky down <controller_name>``. See `Important considerations`_ below.

.. note::
    While the controller requires Kubernetes for High Availability mode, service replicas can still use any cloud or infrastructure supported by SkyPilot.

> How it works

Reviewer 2: It might be good to have a small diagram in this section.

The entire recovery process is handled transparently by SkyPilot and Kubernetes, requiring no action from users when failures occur.

Configuration details
---------------------
Besides the main ``serve.controller.high_availability: true`` flag, you can customize High Availability behavior further:

.. raw:: html

   <ul>
   <li><strong>Controller Resources (<code>serve.controller.resources</code>):</strong> As usual, you can specify <code>cloud</code> (must be Kubernetes), <code>region</code>, <code>cpus</code>, etc. The <code>disk_size</code> here directly determines the size of the PersistentVolumeClaim created for the High Availability controller.</li>
   <li><strong>Kubernetes Storage Class (<code>kubernetes.high_availability.storage_class_name</code> - Optional):</strong> If your Kubernetes cluster has specific storage classes defined (e.g., for different performance tiers like SSD vs HDD, or specific features like backup), you can specify which one to use for the controller's PVC. This is configured under the <code>kubernetes</code> section in <code>config.yaml</code>:</li>

>    </ul>

Reviewer 2: Hmm, why do we use raw HTML here?

Author: Since the code block nested in bold block won't be rendered. https://stackoverflow.com/questions/26366552/bold-code-in-rst

Author: Or let's just skip stressing this `kubernetes.high_availability.storage_class_name`, and mention it in the normal text?

.. code-block:: yaml

    kubernetes:
      # ... other kubernetes settings ...
      high_availability:
        # Optional: Specify the StorageClass name for the controller's PVC
        storage_class_name: <your-storage-class-name> # e.g., premium-ssd

> **Purpose:** Different storage classes offer varying performance (IOPS, throughput), features (snapshots, backups), and costs. If your cluster provides multiple options and you have specific requirements for the controller's storage (e.g., needing faster disk I/O or a particular backup strategy), you can specify a storage class. If omitted, the default storage class configured in your Kubernetes cluster will be used.

Reviewer 2: Can be put in a note box

* **Inconsistent State Error:** If you attempt to launch a service (``sky serve up``) and the ``high_availability`` setting in your ``config.yaml`` *conflicts* with the actual state of the existing SkyServe controller cluster on Kubernetes (e.g., you enabled High Availability in config, but the controller exists as a non-High Availability Pod, or vice-versa), SkyPilot will raise an ``InconsistentHighAvailabilityError``. To resolve this, terminate all services, tear down the controller (``sky down <sky-serve-controller-name>``), and then run ``sky serve up`` again with the desired consistent configuration.

Recovery example
----------------
This example demonstrates the automatic recovery capability of the High Availability controller:

> **0. Preparatory Steps (Ensure Clean State & Correct Config):**

Reviewer 2: Let's start with 1

How it works
------------
When ``high_availability: true`` is set, SkyPilot modifies how the SkyServe controller is deployed on Kubernetes:

1.  **Kubernetes Deployment:** Instead of launching a single Kubernetes Pod, the controller is launched as a Kubernetes `Deployment <https://kubernetes.io/docs/concepts/workloads/controllers/deployment/>`_ with ``replicas: 1``. The Deployment ensures that one instance of the controller pod is always running. If the pod crashes or the node it's on fails, Kubernetes automatically reschedules and starts a new pod.
2.  **Persistent Volume Claim (PVC):** Controller state, including the service database (SQLite), logs, and potentially other runtime information, needs to persist across pod restarts. SkyPilot automatically creates a `PersistentVolumeClaim (PVC) <https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims>`_ for the controller. This PVC is mounted into the controller pod (typically at ``/home/sky``, which includes ``~/.sky/``). The size of the PVC is determined by the ``disk_size`` specified in the controller's ``resources`` configuration (or the default if not specified).

> 3.  **Restart Policy:** The controller pod within the Deployment is configured with ``restartPolicy: Always``.

Reviewer 3: this bullet point feels too detailed and not sure what is the message we want to express.

> 4.  **Persistent Scripts & Recovery:** Setup commands and task ``run`` commands (like the one starting the Python service controller/load balancer) need to be re-executed reliably after a pod restart.

Reviewer 3: also this one. do we really need this?

Recovery example
----------------
This example demonstrates the automatic recovery capability of the HA controller:

**0. Preparatory Steps (Ensure Clean State & Correct Config):**

* **Terminate Existing Controller (if any):**
    * First, ensure **no services are running**. Terminate them with ``sky serve down <service_name>`` or ``sky serve down --all``.
    * Find the controller name:

      .. code-block:: bash

          sky status | grep sky-serve-controller

    * Terminate and purge the controller (replace ``<sky-serve-controller-name>`` with the name you found above):

      .. code-block:: bash

          sky down <sky-serve-controller-name>

* **Set Configuration:** First, ensure your ``~/.sky/config.yaml`` enables HA mode as shown in the `How to enable HA mode`_ section.

  .. code-block:: yaml
      :caption: ~/.sky/config.yaml (relevant part)

      serve:
        controller:
          resources:
            cloud: kubernetes
          high_availability: true

* **Prepare Service Definition:**

  .. code-block:: bash

>       sky api stop

Reviewer 3: why do we need to restart the api server?

        # Should see the default HTML output from http.server

4.  **Simulate Controller Failure (Manually Delete Pod):**
    
    * Find the name of the controller pod. Controller pods typically contain "sky-serve-controller" and have the label ``skypilot-head-node=1``.

    .. code-block:: bash

        kubectl get pods -l skypilot-head-node=1 | grep sky-serve-controller
        # Copy the controller pod name (e.g., sky-serve-controller-deployment-xxxxx-yyyyy)

        CONTROLLER_POD=<paste_controller_pod_name_here>

    * Delete the controller pod.

      .. code-block:: bash

          echo "Deleting controller pod: $CONTROLLER_POD"
          kubectl delete pod $CONTROLLER_POD

5.  **Observe Recovery:** The Kubernetes Deployment will detect the missing pod and automatically create a new one to replace it.

    .. code-block:: bash

        echo "Waiting for controller pod to recover..."
        # Wait a few seconds for Kubernetes to react
        sleep 15

        # Check that a new pod has started and is running (Status should be Running 1/1)
        kubectl get pods -l skypilot-head-node=1
        # Note the pod name will be different, and STATUS should be Running

> 6.  **Verify Service Again:** Even though the controller pod was restarted, the service endpoint should still be accessible (there might be a brief interruption depending on load balancer and K8s response times).

Reviewer 3: ```suggestion
6.  **Verify Service Again:** Even though the controller pod was restarted, the service endpoint should remains the same and still be accessible (there might be a brief interruption depending on load balancer and K8s response times).
```

          cpus: 2+           # Optional, example value
          # disk_size: 100     # Optional, example value (affects PVC size in High Availability mode)

        # --- Enable High Availability ---
        high_availability: true

.. note::
    Enabling or disabling ``high_availability`` only affects **new** SkyServe controllers. If you have an existing controller (either running or stopped), changing this setting will not modify it. To apply the change, you must first terminate all services and then tear down the existing controller using ``sky down <controller_name>``. See `Important considerations`_ below.

.. note::
    While the controller requires Kubernetes for High Availability mode, service replicas can still use any cloud or infrastructure supported by SkyPilot.

How it works
------------
When ``high_availability: true`` is set, SkyPilot modifies how the SkyServe controller is deployed on Kubernetes:

The High Availability implementation relies on standard Kubernetes mechanisms to ensure controller resilience:

* **Automatic Recovery:** The controller runs as a Kubernetes Deployment that automatically restarts pods after failures.

> * **Persistent State:** Critical controller state (database, configuration) is stored on persistent storage that survives pod restarts.

Reviewer 3: ```suggestion
* **Persistent State:** Critical controller state (database, configuration) is stored on persistent storage that persists on pod restarts.
```

> * **Seamless Continuation:** When a new pod starts after a failure, it automatically reconnects to existing resources and continues operations without manual intervention.

Reviewer 3: IIUC for now the load balancer will still down for a while. we should document this behaviour, and add a hint on decoupling lb and controller is wip.

Author: Do you have a PR or issue that tracking the status for this?
